ltv_mpc/mpc_qp.py

```diff
@@ -0,0 +1,165 @@
+            raise ProblemDefinitionError("initial state is undefined")
+        initial_state: np.ndarray = mpc_problem.initial_state
+
+        phi = np.eye(state_dim)
+        psi = np.zeros((state_dim, stacked_input_dim))
+        G_list, h_list = [], []
+        phi_list, psi_list = [], []
+        for k in range(mpc_problem.nb_timesteps):
+            # Loop invariant: x == psi * U + phi * x_init
+            phi_list.append(phi)
+            psi_list.append(psi)
+            A_k = mpc_problem.get_transition_state_matrix(k)
+            B_k = mpc_problem.get_transition_input_matrix(k)
+            C_k = mpc_problem.get_ineq_state_matrix(k)
+            D_k = mpc_problem.get_ineq_input_matrix(k)
+            e_k = mpc_problem.get_ineq_vector(k)
+            G_k = np.zeros((e_k.shape[0], stacked_input_dim))
+            h_k = (
+                e_k
+                if C_k is None
+                else e_k - np.dot(C_k.dot(phi), initial_state)
+            )
+            input_slice = slice(k * input_dim, (k + 1) * input_dim)
+            if D_k is not None:
+                # we rely on G == 0 to avoid a slower +=
+                G_k[:, input_slice] = D_k
+            if C_k is not None:
+                G_k += C_k.dot(psi)
+            if k == 0 and D_k is None and np.any(h_k < 0.0):
+                # in this case, the initial state constraint is violated and
+                # cannot be compensated by any input (D_k is None)
+                warn(
+                    "initial state is unfeasible: "
+                    f"G_0 * x <= h_0 with G_0 == 0 and min(h_0) == {min(h_k)}"
+                )
+            G_list.append(G_k)
+            h_list.append(h_k)
+            phi = A_k.dot(phi)
+            psi = A_k.dot(psi)
+            psi[:, input_slice] = B_k
+        G: np.ndarray = np.vstack(G_list)
+        h: np.ndarray = np.hstack(h_list)
+        Phi = np.vstack(phi_list)
+        Psi = np.vstack(psi_list)
+
+        P: np.ndarray = mpc_problem.stage_input_cost_weight * np.eye(
+            stacked_input_dim,
+        )
+        q: np.ndarray = np.zeros(stacked_input_dim)
+        if mpc_problem.has_terminal_cost:
+            if mpc_problem.goal_state is None:
+                raise ProblemDefinitionError("goal state is undefined")
+            P += mpc_problem.terminal_cost_weight * np.dot(psi.T, psi)
+        if mpc_problem.has_stage_state_cost:
+            if mpc_problem.target_states is None:
+                raise ProblemDefinitionError(
+                    "reference trajectory is undefined"
+                )
+            P += mpc_problem.stage_state_cost_weight * np.dot(Psi.T, Psi)
+
+        self.P = csc_matrix(P) if sparse else P
+        self.G = csc_matrix(G) if sparse else G
+        self.Phi = Phi
+        self.Psi = Psi
+        self.h = h
+        self.phi_last = phi
+        self.psi_last = psi
+        self.q = q  # initialized below
+        #
+        self.update_cost_vector(mpc_problem)
+
```

Comment:
I am wondering if here there is a call missing to `update_constraint_vector(mpc_problem)`.

Reply:
okay, my bad. I saw that for now, the implementation of update_constraint_vector is not ready.